scene/resources/material.cpp

```diff
@@ -700,252 +700,298 @@ void BaseMaterial3D::_update_shader() {
+			code += ", alpha_to_coverage_and_one";
 		}
 	}
 
-	code += ";\n";
+	code += R"(;
+
+uniform vec4 albedo : source_color;
+uniform sampler2D texture_albedo : source_color, )" +
+			texfilter_str + ";\n";
 
-	code += "uniform vec4 albedo : source_color;\n";
-	code += "uniform sampler2D texture_albedo : source_color," + texfilter_str + ";\n";
 	if (grow_enabled) {
-		code += "uniform float grow;\n";
+		code += "uniform float grow : hint_range(-16.0, 16.0, 0.001);\n";
 	}
 
 	if (proximity_fade_enabled) {
-		code += "uniform float proximity_fade_distance;\n";
+		code += "uniform float proximity_fade_distance : hint_range(0.0, 4096.0, 0.01);\n";
 	}
 	if (distance_fade != DISTANCE_FADE_DISABLED) {
-		code += "uniform float distance_fade_min;\n";
-		code += "uniform float distance_fade_max;\n";
+		code += R"(
+uniform float distance_fade_min : hint_range(0.0, 4096.0, 0.01);
+uniform float distance_fade_max : hint_range(0.0, 4096.0, 0.01);
+)";
 	}
 
 	if (flags[FLAG_ALBEDO_TEXTURE_MSDF]) {
-		code += "uniform float msdf_pixel_range;\n";
-		code += "uniform float msdf_outline_size;\n";
+		code += R"(
+uniform float msdf_pixel_range : hint_range(1.0, 100.0, 1.0);
+uniform float msdf_outline_size : hint_range(0.0, 250.0, 1.0);
+)";
 	}
 
-	// alpha scissor is only valid if there is not antialiasing edge
-	// alpha hash is valid whenever, but not with alpha scissor
+	// Alpha scissor is only valid if there is no antialiasing edge.
+	// Alpha hash is valid whenever, but not with alpha scissor.
 	if (transparency == TRANSPARENCY_ALPHA_SCISSOR) {
-		code += "uniform float alpha_scissor_threshold;\n";
+		code += "uniform float alpha_scissor_threshold : hint_range(0.0, 1.0, 0.001);\n";
 	} else if (transparency == TRANSPARENCY_ALPHA_HASH) {
-		code += "uniform float alpha_hash_scale;\n";
+		code += "uniform float alpha_hash_scale : hint_range(0.0, 2.0, 0.01);\n";
 	}
-	// if alpha antialiasing isn't off, add in the edge variable
+	// If alpha antialiasing isn't off, add in the edge variable.
 	if (alpha_antialiasing_mode != ALPHA_ANTIALIASING_OFF &&
 			(transparency == TRANSPARENCY_ALPHA_SCISSOR || transparency == TRANSPARENCY_ALPHA_HASH)) {
-		code += "uniform float alpha_antialiasing_edge;\n";
-		code += "uniform ivec2 albedo_texture_size;\n";
+		code += R"(
+uniform float alpha_antialiasing_edge : hint_range(0.0, 1.0, 0.01);
+uniform ivec2 albedo_texture_size;
+)";
 	}
 
-	code += "uniform float point_size : hint_range(0,128);\n";
+	code += "uniform float point_size : hint_range(0.1, 128.0, 0.1);\n";
 
-	//TODO ALL HINTS
 	if (!orm) {
-		code += "uniform float roughness : hint_range(0,1);\n";
-		code += "uniform sampler2D texture_metallic : hint_default_white," + texfilter_str + ";\n";
-		code += "uniform vec4 metallic_texture_channel;\n";
+		code += R"(
+uniform float roughness : hint_range(0.0, 1.0);
+uniform sampler2D texture_metallic : hint_default_white, )" +
+				texfilter_str + R"(;
+uniform vec4 metallic_texture_channel;
+)";
```

Comment:
Likewise, if this can work with `vformat`, I think it would be more readable. The closing parenthesis of the raw-string format in the middle of the string really hurts readability IMO, in something that can typically be generated code with parentheses.

Reply:
For me, for raw strings to be usable, there needs to be a clear separation between the `R"(` starting and closing `)"` tokens, and the actually code, so never use `R"(;` to start with a closing semicolon, and avoid concatenation and re-starting a raw-string mid statement.

```suggestion
		code += vformat(R"(
uniform float roughness : hint_range(0.0, 1.0);
uniform sampler2D texture_metallic : hint_default_white, %s;
uniform vec4 metallic_texture_channel;
)",
				texfilter_str);
```

Reply:
You missed changing this one it seems.